lib/editor-linter.coffee

```diff
@@ -1,102 +1,41 @@
-{CompositeDisposable, Emitter, Range} = require 'atom'
-Helpers = require './helpers'
+{TextEditor, Emitter, CompositeDisposable} = require('atom')
 
 class EditorLinter
-  constructor: (@linter, @editor) ->
-    @status = true
-    @messages = new Map
-    @inProgress = false
-    @inProgressFly = false
-
-    if @editor is atom.workspace.getActiveTextEditor()
-      @linter.views.updateLineMessages(true)
-
+  constructor: (@editor) ->
+    throw new Error("Given editor isn't really an editor") unless @editor instanceof TextEditor
     @emitter = new Emitter
     @subscriptions = new CompositeDisposable
-
-    @subscriptions.add(
-      @editor.onDidSave => @lint(false)
-    )
-    @subscriptions.add(
-      @editor.onDidChangeCursorPosition ({oldBufferPosition, newBufferPosition}) =>
+    @subscriptions.add @editor.onDidDestroy =>
+      @emitter.emit 'did-destroy'
+
+    @subscriptions.add @editor.onDidSave => @emitter.emit('should-lint', false)
+    # The onDidStopChanging callbacks are invoked immediately, and we want to avoid it
+    setImmediate =>
```

Comment:
I see your comment and I still don't follow how it's needed.

`editorLinter = new EditorLinter(editor)` When this code executes the outer `setImmediate` is called. The rest of the `atom.workspace.observeTextEditors` callback completes, then we come back here and subscribe to `@editor.onDidStopChanging`. It calls the callback immediately, which triggers a 2nd setImmediate. At that point the rest of this function gets to complete (`@editor.onDidChangeCursorPosition`). After it completes then the callback executes that triggers `@emitter.emit('should-lint', true)`

So why can't adding these subscriptions happen during the construction of the class? the `should-lint` event is already asynchronous